### PR TITLE
bpo-32759: Free unused arenas in multiprocessing.heap

### PR DESCRIPTION
**Pull request by pitrou**
Large shared arrays allocated using multiprocessing would remain allocated
until the process ends.

https://bugs.python.org/issue32759

### CODE REVIEW

File: Lib/multiprocessing/heap.py [hunk 1]
```diff
@@ -237,14 +277,15 @@ def malloc(self, size):
...
         with self._lock:
+            self._n_mallocs += 1
             self._free_pending_blocks()
             size = self._roundup(max(size,1), self._alignment)
             (arena, start, stop) = self._malloc(size)
             new_stop = start + size
             if new_stop < stop:
-                self._free((arena, new_stop, stop))
+                self._add_free_block((arena, new_stop, stop))
             block = (arena, start, new_stop)
```
Comment (serhiy-storchaka) [inline Lib/multiprocessing/heap.py]: The `block` variable can be inlined now.

**Review by serhiy-storchaka — COMMENTED** [review]
I started reviewing yesterday, but it will take some time to grok this code.

File: Lib/multiprocessing/heap.py (hunk 1 above)
Comment (pitrou) [reply Lib/multiprocessing/heap.py]: Indeed :-)

**Comment by pitrou** [comment]
> but it will take some time to grok this code.

Yes, it's a mini-memory allocator here. Not too complex fortunately.

**Comment by pitrou** [comment]
I adress the review comment and added some comments to try and explain the allocator a bit more.

If there are no further review comments, I will merge this PR.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**